File: Workspace/src/talker_listern_py/talker_listern_py/keyboard_talker_char.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import sys
import termios
import tty

class Talker(Node):
    def __init__(self):
        super().__init__('talker')
        self.publisher_ = self.create_publisher(String, 'chatter', 10)
        self.orig_termios = termios.tcgetattr(sys.stdin)
        tty.setraw(sys.stdin)

    def __del__(self):
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, self.orig_termios)

    def capture_and_publish(self):
        try:
            while rclpy.ok():
                c = sys.stdin.read(1)
                if c == '\x03':  # Ctrl-C to exit
                    break
                msg = String()
                msg.data = c
                self.get_logger().info(f'Publishing: "{msg.data}"')
                self.publisher_.publish(msg)
        except Exception as e:
            self.get_logger().error(f'Exception: {e}')
        finally:
            termios.tcsetattr(sys.stdin, termios.TCSADRAIN, self.orig_termios)

def main(args=None):
    rclpy.init(args=args)
    node = Talker()
    try:
        node.capture_and_publish()
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Workspace/src/talker_listern_cpp/keyboard_talker_char.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include <chrono>
#include <iostream>
#include <memory>
#include <termios.h>
#include <unistd.h>


class Talker : public rclcpp::Node {
public:
  Talker() : Node("talker") {
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
    // Set terminal to raw mode to capture keyboard input
    tcgetattr(STDIN_FILENO, &orig_termios_);
    termios raw = orig_termios_;
    raw.c_lflag &= ~(ECHO | ICANON);
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw);
  }

  ~Talker() {
    // Restore terminal settings
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &orig_termios_);
  }

  void capture_and_publish() {
    char c;
    while (rclcpp::ok()) {
      c = getchar();
      if (c == '\x03') { // Ctrl-C to exit
        break;
      }
      auto message = std_msgs::msg::String();
      message.data = std::string(1, c);
      RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
      publisher_->publish(message);
    }
  }

private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  termios orig_termios_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<Talker>();
  std::thread capture_thread(&Talker::capture_and_publish, node);
  rclcpp::spin(node);
  capture_thread.join();
  rclcpp::shutdown();
  return 0;
}
